csrc/device_lower/pass/grid_serialization.cpp

```diff
@@ -53,45 +54,156 @@ class GridSerializationSyncInserter : kir::ExprMutator {
 
-      if (cur_expr_sync_pattern_.has_value()) {
-        NVF_ERROR(
-            cur_expr_sync_pattern_.value() == sync_pattern,
-            "Reduction op ",
-            rop->toString(),
-            " has requested serial grid reduction, but pattern ",
-            sync_pattern.toString(),
-            " conflicts with previous pattern: ",
-            cur_expr_sync_pattern_.value().toString());
-      } else {
-        cur_expr_sync_pattern_ = sync_pattern;
+    Val* is_first_step = nullptr;
+    Val* is_last_step = nullptr;
+    for (auto pt : kParallelTypeBIDs) {
+      if (cur_expr_sync_pattern_.value().get(pt)) {
+        // && BID == 0
+        is_first_step = SimplifyingIrBuilder::logicalAndExpr(
+            is_first_step,
+            IrBuilder::eqExpr(
+                NamedScalar::getParallelIndex(pt),
+                FusionGuard::getCurFusion()->zeroVal(DataType::Index)));
+        // && BID == BDIM - 1
+        is_last_step = SimplifyingIrBuilder::logicalAndExpr(
+            is_last_step,
+            IrBuilder::eqExpr(
+                NamedScalar::getParallelIndex(pt),
+                IrBuilder::subExpr(
+                    NamedScalar::getParallelDim(pt),
+                    FusionGuard::getCurFusion()->oneVal(DataType::Index))));
       }
     }
+    is_first_step = GpuLower::current()->commonScalarMap().hoistScalar(
+        is_first_step, for_loops_);
+    is_last_step = GpuLower::current()->commonScalarMap().hoistScalar(
+        is_last_step, for_loops_);
+    NVF_ERROR(is_first_step != nullptr);
+    NVF_ERROR(is_last_step != nullptr);
+
+    auto in = rop->in()->as<TensorView>();
+    auto out = rop->out()->as<TensorView>();
+
+    // Create a work buffer
+    TensorView* work_buffer_in = IrBuilder::create<TensorView>(
+        out->domain(), out->dtype(), MemoryType::Global);
+
+    TensorView* loaded_work_buffer =
+        work_buffer_in->cacheAfter(LoadStoreOpType::Set, CacheOp::Global);
+    // INTERNAL ASSERT FAILED at
+    // "/opt/pytorch/nvfuser/csrc/tensor_view.cpp":1278 Function invalid for
+    // kernel container.
+    // TODO: caching is not meant to work on kernel IR, so we'll have to work
+    // around that
+
+    TensorView* in_plus_work =
+        binaryOp(rop->getReductionOpType(), in, loaded_work_buffer);
+    TensorView* result = where(is_first_step, in, in_plus_work);
+
+    // work_buffer_out is dead code (it has no uses). This means we won't
+    // automatically insert new Allocate nodes for it in subsequent lowering
+    // passes, so it is safe for us to create one here.
+    TensorView* work_buffer_out =
+        result->cacheAfter(LoadStoreOpType::Set, CacheOp::Global);
+    work_buffer_out->setMemoryType(MemoryType::Global);
+    // TODO: we cannot force aliasing work_buffer_out => work_buffer_in yet,
+    // since that is done with Allocate. Since we _do_ want automatic insertion
+    // of work_buffer_in's Allocate, it is unclear how we can effectively set
+    // this alias.
```

Comment:
@naoyam I think this is the biggest challenge to using IR to represent `serialReductionStep`. We could try and manually set that alias in a later pass, but once we have done this translation, it might be difficult to recognize this as a reduction step.